libcaf_net/caf/net/datagram_transport.hpp

```diff
@@ -19,193 +19,118 @@
-    if (auto err = dispatcher_.init(*this))
+  error init(endpoint_manager& manager) override {
+    CAF_LOG_TRACE("");
+    if (auto err = super::init(manager))
       return err;
+    prepare_next_read();
     return none;
   }
 
-  template <class Parent>
-  bool handle_read_event(Parent&) {
-    CAF_LOG_TRACE(CAF_ARG(handle_.id));
-    auto ret = read(handle_, make_span(read_buf_));
+  bool handle_read_event(endpoint_manager&) override {
+    CAF_LOG_TRACE(CAF_ARG(this->handle_.id));
+    auto ret = read(this->handle_, make_span(this->read_buf_));
     if (auto res = get_if<std::pair<size_t, ip_endpoint>>(&ret)) {
       auto num_bytes = res->first;
+      CAF_LOG_DEBUG("received " << num_bytes << " bytes");
       auto ep = res->second;
-      read_buf_.resize(num_bytes);
-      dispatcher_.handle_data(*this, make_span(read_buf_), std::move(ep));
+      this->read_buf_.resize(num_bytes);
+      this->next_layer_.handle_data(*this, make_span(this->read_buf_),
+                                    std::move(ep));
       prepare_next_read();
     } else {
       auto err = get<sec>(ret);
       CAF_LOG_DEBUG("send failed" << CAF_ARG(err));
-      dispatcher_.handle_error(err);
+      this->next_layer_.handle_error(err);
       return false;
     }
     return true;
   }
 
-  template <class Parent>
-  bool handle_write_event(Parent& parent) {
-    CAF_LOG_TRACE(CAF_ARG(handle_.id)
+  bool handle_write_event(endpoint_manager& manager) override {
+    CAF_LOG_TRACE(CAF_ARG(this->handle_.id)
                   << CAF_ARG2("queue-size", packet_queue_.size()));
     // Try to write leftover data.
     write_some();
     // Get new data from parent.
-    for (auto msg = parent.next_message(); msg != nullptr;
-         msg = parent.next_message()) {
-      dispatcher_.write_message(*this, std::move(msg));
+    for (auto msg = manager.next_message(); msg != nullptr;
+         msg = manager.next_message()) {
+      this->next_layer_.write_message(*this, std::move(msg));
     }
     // Write prepared data.
     return write_some();
   }
 
-  template <class Parent>
-  void resolve(Parent&, const uri& locator, const actor& listener) {
-    dispatcher_.resolve(*this, locator, listener);
-  }
-
-  template <class Parent>
-  void new_proxy(Parent&, const node_id& peer, actor_id id) {
-    dispatcher_.new_proxy(*this, peer, id);
-  }
-
-  template <class Parent>
-  void local_actor_down(Parent&, const node_id& peer, actor_id id,
-                        error reason) {
-    dispatcher_.local_actor_down(*this, peer, id, std::move(reason));
-  }
-
-  template <class Parent>
-  void timeout(Parent&, atom_value value, uint64_t id) {
-    dispatcher_.timeout(*this, value, id);
-  }
-
-  void set_timeout(uint64_t timeout_id, id_type id) {
-    dispatcher_.set_timeout(timeout_id, id);
-  }
-
-  void handle_error(sec code) {
-    dispatcher_.handle_error(code);
-  }
-
+  // TODO: remove this function. `resolve` should add workers when needed.
```

Comment:
Hm?

Reply:
Adding new workers for new connections shouldn't happen manually. If a resolve_request succeeds a new `transport_worker` should be added automatically. This is a function needed for the current test implementation because `resolve` doesn't actually create workers automatically by now.. I will implement that and remove this function later on.